packages/vulcan-ui-bootstrap/lib/components/forms/FormComponentInner.jsx

```diff
@@ -0,0 +1,208 @@
+import React, { PureComponent } from 'react';
+import PropTypes from 'prop-types';
+import { intlShape } from 'meteor/vulcan:i18n';
+import { Components, registerComponent, instantiateComponent } from 'meteor/vulcan:core';
+import classNames from 'classnames';
+
+
+class FormComponentInner extends PureComponent {
+  
+  renderClear = () => {
+    if (['datetime', 'time', 'select', 'radiogroup'].includes(this.props.input)) {
+      return (
+        <a
+          href="javascript:void(0)"
+          className="form-component-clear"
+          title={this.context.intl.formatMessage({ id: 'forms.clear_field' })}
+          onClick={this.props.clearField}
+        >
+          <span>✕</span>
+        </a>
+      );
+    }
+  };
+  
+  getProperties = () => {
+    const { name, options, label, onChange, value } = this.props;
+  
+    // these properties are whitelisted so that they can be safely passed to the actual form input
+    // and avoid https://facebook.github.io/react/warnings/unknown-prop.html warnings
+    const inputProperties = {
+      name,
+      options,
+      label,
+      onChange,
+      value,
+      ...this.props.inputProperties,
+    };
+  
+    return {
+      ...this.props,
+      inputProperties,
+    };
+  };
+  
+  renderExtraComponent = (extraComponent) => {
+    if (!extraComponent) return null;
+  
+    const properties = this.getProperties();
+    
+    return instantiateComponent(extraComponent, properties);
+  };
+  
+  
+  renderComponent = () => {
```

Comment:
I thought this would stay in `FormComponent` and we would just pass down the component as a prop. I think `FormComponentInner` should stay as light as possible, if it's going to be duplicated between each various UI package.

Reply:
Let me look at this with a fresh mind tomorrow and see what I can do about that.

Reply:
Sounds good :)

Reply:
Looks like I can accomplish what you are looking for. I do like having more in the shared `FormComponent` and as little as possible in the multiple `FormComponentInner`s, as long as each UI has all the control it needs to render in their native way.